sopel/tools/jobs.py

```diff
@@ -127,75 +199,284 @@ def _call(self, job):
-    many times at once.
     """
+    @classmethod
+    def kwargs_from_callable(cls, handler):
+        """Generate the keyword arguments to create a new instance.
+
+        :param callable handler: callable used to generate keyword arguments
+        :return: a map of keyword arguments
+        :rtype: dict
+
+        This classmethod takes the ``handler``'s attributes to generate a map
+        of keyword arguments for the class. This can be used by the
+        :meth:`from_callable` classmethod to instantiate a new rule object.
 
-    def __init__(self, interval, func):
-        self.next_time = time.time() + interval
-        self.interval = interval
-        self.func = func
+        The expected attributes are the ones set by decorators from the
+        :mod:`sopel.module` module.
+        """
+
+        return {
+            'plugin': getattr(handler, 'plugin_name', None),
+            'label': getattr(handler, 'rule_label', None),
+            'threaded': getattr(handler, 'thread', True),
+            'doc': inspect.getdoc(handler),
+        }
+
+    @classmethod
+    def from_callable(cls, settings, handler):
+        """Instantiate a Job from the bot's ``settings`` and a ``handler``.
+
+        :param settings: bot's settings
+        :type settings: :class:`sopel.config.Config`
+        :param handler: callable used to instantiate a new job
+        :type handler: :term:`function`
+        """
+        kwargs = cls.kwargs_from_callable(handler)
+        return cls(
+            set(handler.interval),
+            handler=handler,
+            **kwargs)
+
+    def __init__(self,
+                 intervals,
+                 plugin=None,
+                 label=None,
+                 handler=None,
+                 threaded=True,
+                 doc=None):
+        # scheduling
+        now = time.time()
+        self.intervals = set(intervals)
+        """Set of intervals to execute the job at."""
+        self.next_times = dict(
+            (interval, now + interval)
+            for interval in self.intervals
+        )
+        """Tracking of when to execute the job next time."""
+
+        # meta
+        self._plugin_name = plugin
+        self._label = label
+        self._doc = doc
+
+        # execution
+        self._handler = handler
+        self._threaded = bool(threaded)
+        self.is_running = threading.Event()
+        """Running flag: it tells if the job is running or not.
+
+        This flag is set and clear automatically by the :meth:`execute` method.
+        It is also set and clear when the job is used with the ``with``
+        statement::
+
+            with job:
+                # you do something before executing the job
+                # this ensure that the job is marked as "running"
+
+        .. note::
+
+            When set manually or with the ``with`` statement, the
+            :meth:`execute` won't clear this attribute.
```

Comment:
```suggestion
            :meth:`execute` method won't clear this attribute itself.
```

Still make sense / match what you meant?

Reply:
Actually it's even better.